{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from brian2 import *\n",
    "%matplotlib inline\n",
    "\n",
    "#basic neuronal group characteristics\n",
    "neurons_num = 12 #referring to the number of stuctures and subgroups sizes\n",
    "\n",
    "v_rest = -70*mV #resting potential\n",
    "v_th = -50*mV #threshold potential\n",
    "area = 0.785*umetre**2\n",
    "Cm = 1*ufarad*cm**-2 * area\n",
    "v_eK = -90*mV\n",
    "v_eNa = 50*mV\n",
    "g_Na = 100*msiemens*cm**-2 * area\n",
    "g_K = 30*msiemens*cm**-2 * area\n",
    "\n",
    "\n",
    "#Hodgkin-Huxley model equations\n",
    "eqs = '''\n",
    "dv/dt = -(g_Na*(m*m*m)*h*(v-v_eNa) + g_K*(n*n*n*n)*(v-v_eK) + I_stim)/Cm : volt\n",
    "dn/dt = ((0.01*(10 + v))/(exp((10 + v)/10) - 1))*(1-n) - (0.125*exp(v/80))*n : 1\n",
    "dm/dt = ((0.1*(25 + v))/(exp((25 + v)/10) - 1))*(1-m) - (4*exp(v/18))*m : 1\n",
    "dh/dt = 0.07*exp(v/20)*(1-h) - (1/(exp((30 + v)/10) + 1))*h : 1\n",
    "I_stim : amp\n",
    "\n",
    "'''\n",
    "\n",
    "#glutamate\n",
    "model_eqs_glu = '''\n",
    "receptors_num : 1\n",
    "v_e : volt\n",
    "W : metre**3\n",
    "Kd : mM\n",
    "gam : siemens\n",
    "dL/dt = -1.1*10**-9 : mol (event-driven)\n",
    "dI_stim_syn/dt = receptors_num*(v - v_e)*gam*((L/W/Kd)/(1 + (L/W/Kd))) : amp (clock-driven)\n",
    "\n",
    "'''\n",
    "\n",
    "#gaba\n",
    "model_eqs_gaba = '''\n",
    "receptors_num : 1\n",
    "v_e : volt\n",
    "W : metre**3\n",
    "Kd : mM\n",
    "gam : siemens\n",
    "dL/dt = -4*10**-9 : mol (event-driven)\n",
    "dI_stim_syn/dt = receptors_num*(v - v_e)*gam*((L/W/Kd)/(1 + (L/W/Kd))) : amp (clock-driven)\n",
    "\n",
    "'''\n",
    "\n",
    "Prototype = NeuronGroup(neurons_num, eqs, threshold='v > v_th')\n",
    "Prototype.v = v_rest\n",
    "\n",
    "# Mapping subgroups by cerebellum integration circuit\n",
    "MC_in = Prototype[0]\n",
    "DSGo = Prototype[1]\n",
    "DSNoGo = Prototype[2]\n",
    "GPi = Prototype[3]\n",
    "GPe = Prototype[5]\n",
    "STN = Prototype[6]\n",
    "Th = Prototype[4]\n",
    "MC_out = Prototype[11]\n",
    "\n",
    "Pons = Prototype[7]\n",
    "CC = Prototype[8]\n",
    "DN = Prototype[9]\n",
    "\n",
    "SNc = Prototype[10]\n",
    "#Glu = Prototype[1, 2, 4, 7, 8, 11]\n",
    "#GABA = Prototype[3, 4, 5, 9]\n",
    "\n",
    "#Setting up synapses\n",
    "#excitatory synapses (glutamate)\n",
    "S_exc = Synapses(Prototype, Prototype, model = Equations(model_eqs_glu), on_pre='L = 1.1*mM')\n",
    "\n",
    "\n",
    "S_exc.connect(i=MC_in, j=[1, 2])\n",
    "S_exc.connect(i=Th, j=MC_out)\n",
    "#S_exc.connect(i=SNc, j=DSGo)\n",
    "\n",
    "#cerebellum route\n",
    "S_exc.connect(i=MC_in, j=Pons)\n",
    "S_exc.connect(i=Pons, j=CC)\n",
    "S_exc.connect(i=DN, j=Th)\n",
    "\n",
    "#synapses parameters\n",
    "S_exc.receptors_num = 2500\n",
    "S_exc.v_e = +55*mV\n",
    "S_exc.gam = 10*psiemens\n",
    "S_exc.W = 20*nmeter * area\n",
    "S_exc.Kd = 500*nM\n",
    "\n",
    "\n",
    "#inhibitory synapses (GABA)\n",
    "S_inh = Synapses(Prototype, Prototype, model = Equations(model_eqs_gaba), on_pre='L = 4*mM')\n",
    "                \n",
    "S_inh.connect(i=DSGo, j=GPi)\n",
    "S_inh.connect(i=DSNoGo, j=GPe)\n",
    "S_inh.connect(i=GPe, j=GPi)\n",
    "S_inh.connect(i=GPi, j=Th)\n",
    "#S_inh.connect(i=SNc, j=DSNoGo)\n",
    "\n",
    "#cerebellum route\n",
    "S_inh.connect(i=CC, j=DN)\n",
    "\n",
    "#synapses parameters\n",
    "S_inh.receptors_num = 2500\n",
    "S_inh.v_e = -65*mV\n",
    "S_inh.gam = 8*psiemens\n",
    "S_inh.W = 20*nmeter * area\n",
    "S_inh.Kd = 128*nM\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "BrianObjectException",
     "evalue": "Original error and traceback:\nTraceback (most recent call last):\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/equations.py\", line 960, in check_units\n    check_dimensions(str(eq.expr), self.dimensions[var] / second.dim,\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/unitcheck.py\", line 41, in check_dimensions\n    expr_dims = parse_expression_dimensions(expression, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 326, in parse_expression_dimensions\n    left_dim = parse_expression_dimensions(expr.left, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 326, in parse_expression_dimensions\n    left_dim = parse_expression_dimensions(expr.left, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 327, in parse_expression_dimensions\n    right_dim = parse_expression_dimensions(expr.right, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 311, in parse_expression_dimensions\n    raise DimensionMismatchError(msg)\nbrian2.units.fundamentalunits.DimensionMismatchError: Argument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/core/network.py\", line 901, in before_run\n    obj.before_run(run_namespace)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/groups/neurongroup.py\", line 888, in before_run\n    self.equations.check_units(self, run_namespace=run_namespace)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/equations.py\", line 963, in check_units\n    raise DimensionMismatchError(('Inconsistent units in '\nbrian2.units.fundamentalunits.DimensionMismatchError: Inconsistent units in differential equation defining variable h:\nArgument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).\n\nError encountered with object named \"neurongroup_3\".\nObject was created here (most recent call only, full details in debug log):\n  File \"<ipython-input-10-7951390132da>\", line 51, in <module>\n    Prototype = NeuronGroup(neurons_num, eqs, threshold='v > v_th')\n\nAn error occurred when preparing an object. brian2.units.fundamentalunits.DimensionMismatchError: Inconsistent units in differential equation defining variable h:\nArgument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).\n(See above for original error message and traceback.)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mDimensionMismatchError\u001b[0m                    Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/equations.py\u001b[0m in \u001b[0;36mcheck_units\u001b[0;34m(self, group, run_namespace)\u001b[0m\n\u001b[1;32m    959\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 960\u001b[0;31m                     check_dimensions(str(eq.expr), self.dimensions[var] / second.dim,\n\u001b[0m\u001b[1;32m    961\u001b[0m                                      all_variables)\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/unitcheck.py\u001b[0m in \u001b[0;36mcheck_dimensions\u001b[0;34m(expression, dimensions, variables)\u001b[0m\n\u001b[1;32m     40\u001b[0m     '''\n\u001b[0;32m---> 41\u001b[0;31m     \u001b[0mexpr_dims\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpression\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m     err_msg = ('Expression {expr} does not have the '\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\u001b[0m in \u001b[0;36mparse_expression_dimensions\u001b[0;34m(expr, variables)\u001b[0m\n\u001b[1;32m    325\u001b[0m         \u001b[0mop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m         \u001b[0mleft_dim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mleft\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m         \u001b[0mright_dim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mright\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\u001b[0m in \u001b[0;36mparse_expression_dimensions\u001b[0;34m(expr, variables)\u001b[0m\n\u001b[1;32m    325\u001b[0m         \u001b[0mop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m         \u001b[0mleft_dim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mleft\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m         \u001b[0mright_dim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mright\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\u001b[0m in \u001b[0;36mparse_expression_dimensions\u001b[0;34m(expr, variables)\u001b[0m\n\u001b[1;32m    326\u001b[0m         \u001b[0mleft_dim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mleft\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 327\u001b[0;31m         \u001b[0mright_dim\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_expression_dimensions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mright\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    328\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mop\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'Add'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Sub'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Mod'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\u001b[0m in \u001b[0;36mparse_expression_dimensions\u001b[0;34m(expr, variables)\u001b[0m\n\u001b[1;32m    310\u001b[0m                         get_dimensions(arg_unit), get_dimensions(expected_unit))\n\u001b[0;32m--> 311\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mDimensionMismatchError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDimensionMismatchError\u001b[0m: Argument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mDimensionMismatchError\u001b[0m                    Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/network.py\u001b[0m in \u001b[0;36mbefore_run\u001b[0;34m(self, run_namespace)\u001b[0m\n\u001b[1;32m    900\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 901\u001b[0;31m                     \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbefore_run\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_namespace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    902\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/groups/neurongroup.py\u001b[0m in \u001b[0;36mbefore_run\u001b[0;34m(self, run_namespace)\u001b[0m\n\u001b[1;32m    887\u001b[0m         \u001b[0;31m# Check units\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 888\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mequations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_units\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_namespace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrun_namespace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    889\u001b[0m         \u001b[0;31m# Check that subexpressions that refer to stateful functions are labeled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/equations.py\u001b[0m in \u001b[0;36mcheck_units\u001b[0;34m(self, group, run_namespace)\u001b[0m\n\u001b[1;32m    962\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mDimensionMismatchError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 963\u001b[0;31m                     raise DimensionMismatchError(('Inconsistent units in '\n\u001b[0m\u001b[1;32m    964\u001b[0m                                                   \u001b[0;34m'differential equation '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDimensionMismatchError\u001b[0m: Inconsistent units in differential equation defining variable h:\nArgument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mBrianObjectException\u001b[0m                      Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-2cef164aa1ea>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mM\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mStateMonitor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPrototype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'v'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecord\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m5000\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mM\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mM\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'Th'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/units/fundamentalunits.py\u001b[0m in \u001b[0;36mnew_f\u001b[0;34m(*args, **kwds)\u001b[0m\n\u001b[1;32m   2390\u001b[0m                                                      get_dimensions(newkeyset[k]))\n\u001b[1;32m   2391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2392\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2393\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;34m'result'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mau\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2394\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mau\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'result'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mbool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/magic.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(duration, report, report_period, namespace, profile, level)\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0mintended\u001b[0m \u001b[0muse\u001b[0m\u001b[0;34m.\u001b[0m \u001b[0mSee\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mMagicNetwork\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mmore\u001b[0m \u001b[0mdetails\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    372\u001b[0m     '''\n\u001b[0;32m--> 373\u001b[0;31m     return magic_network.run(duration, report=report, report_period=report_period,\n\u001b[0m\u001b[1;32m    374\u001b[0m                              namespace=namespace, profile=profile, level=2+level)\n\u001b[1;32m    375\u001b[0m \u001b[0mrun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__module__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/magic.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, duration, report, report_period, namespace, profile, level)\u001b[0m\n\u001b[1;32m    229\u001b[0m             namespace=None, profile=False, level=0):\n\u001b[1;32m    230\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_update_magic_objects\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 231\u001b[0;31m         Network.run(self, duration, report=report, report_period=report_period,\n\u001b[0m\u001b[1;32m    232\u001b[0m                     namespace=namespace, profile=profile, level=level+1)\n\u001b[1;32m    233\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/base.py\u001b[0m in \u001b[0;36mdevice_override_decorated_function\u001b[0;34m(*args, **kwds)\u001b[0m\n\u001b[1;32m    278\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcurdev\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 280\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    281\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    282\u001b[0m         \u001b[0mdevice_override_decorated_function\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/units/fundamentalunits.py\u001b[0m in \u001b[0;36mnew_f\u001b[0;34m(*args, **kwds)\u001b[0m\n\u001b[1;32m   2390\u001b[0m                                                      get_dimensions(newkeyset[k]))\n\u001b[1;32m   2391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2392\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2393\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;34m'result'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mau\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2394\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mau\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'result'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mbool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/network.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, duration, report, report_period, namespace, profile, level)\u001b[0m\n\u001b[1;32m   1010\u001b[0m             \u001b[0mnamespace\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_local_namespace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1011\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1012\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbefore_run\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnamespace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1013\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1014\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_objects\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/base.py\u001b[0m in \u001b[0;36mdevice_override_decorated_function\u001b[0;34m(*args, **kwds)\u001b[0m\n\u001b[1;32m    278\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcurdev\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 280\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    281\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    282\u001b[0m         \u001b[0mdevice_override_decorated_function\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/brian2/core/network.py\u001b[0m in \u001b[0;36mbefore_run\u001b[0;34m(self, run_namespace)\u001b[0m\n\u001b[1;32m    901\u001b[0m                     \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbefore_run\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_namespace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    902\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 903\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mbrian_object_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"An error occurred when preparing an object.\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    904\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    905\u001b[0m         \u001b[0;31m# Check that no object has been run as part of another network before\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mBrianObjectException\u001b[0m: Original error and traceback:\nTraceback (most recent call last):\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/equations.py\", line 960, in check_units\n    check_dimensions(str(eq.expr), self.dimensions[var] / second.dim,\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/unitcheck.py\", line 41, in check_dimensions\n    expr_dims = parse_expression_dimensions(expression, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 326, in parse_expression_dimensions\n    left_dim = parse_expression_dimensions(expr.left, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 326, in parse_expression_dimensions\n    left_dim = parse_expression_dimensions(expr.left, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 327, in parse_expression_dimensions\n    right_dim = parse_expression_dimensions(expr.right, variables)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/parsing/expressions.py\", line 311, in parse_expression_dimensions\n    raise DimensionMismatchError(msg)\nbrian2.units.fundamentalunits.DimensionMismatchError: Argument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/core/network.py\", line 901, in before_run\n    obj.before_run(run_namespace)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/groups/neurongroup.py\", line 888, in before_run\n    self.equations.check_units(self, run_namespace=run_namespace)\n  File \"/Users/victoria.stelmakh/opt/anaconda3/lib/python3.8/site-packages/brian2/equations/equations.py\", line 963, in check_units\n    raise DimensionMismatchError(('Inconsistent units in '\nbrian2.units.fundamentalunits.DimensionMismatchError: Inconsistent units in differential equation defining variable h:\nArgument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).\n\nError encountered with object named \"neurongroup_3\".\nObject was created here (most recent call only, full details in debug log):\n  File \"<ipython-input-10-7951390132da>\", line 51, in <module>\n    Prototype = NeuronGroup(neurons_num, eqs, threshold='v > v_th')\n\nAn error occurred when preparing an object. brian2.units.fundamentalunits.DimensionMismatchError: Inconsistent units in differential equation defining variable h:\nArgument number 1 for function exp does not have the correct units. Expression \"v / 20\" has units (m^2 kg s^-3 A^-1), but should be (1).\n(See above for original error message and traceback.)"
     ]
    }
   ],
   "source": [
    "M = StateMonitor(Prototype, 'v', record=True)\n",
    "\n",
    "run(5000*ms)\n",
    "\n",
    "plot(M.t/ns, M.v[4], label='Th')\n",
    "\n",
    "xlabel('Time (ms)')\n",
    "ylabel('v')\n",
    "legend();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
